### DISCUSSION TITLE
Cross-post navigation between attachments

### DISCUSSION THREAD

**Discussion by indeyets** [FreeFeed/freefeed-react-client#1415] [Ideas]
* Feature name: Cross-post navigation between attachments
* Start Date: 2021-07-31

# Summary

Let users navigate between attachments (mostly images) of different posts and comments without leaving lightbox.

# Motivation

Lightbox lets users see groups of images attached single post or  to single comment. It covers needs of users who use freefeed as a short-messages platform. This should be kept as it is.

It would be nice to support additional use-cases such as:

1. "gallery feed": each post to feed has several attachments and users might want to browse between them without leaving lightbox.
2. "long posts": author creates post with attachments and continues with additional comments which have more attachments
3. "image boards": users interact with each other within a thread using images/memes as the main mean of communication

It is extremely important to make this solution mobile-first. All use-cases must work on mobile phones and tablets. On mobile, user interaction with lightbox is limited to horizontal swiping and on-screen buttons. Vertical-swiping is reserved as it is used for exiting from lightbox. So, the solution should be within this limits.

# User-experience description

I suggest that we add a new "html"-based item to the last position of lightbox galleries, which splits image area with large buttons. This item can be swiped as usual and it will stay as part of the loop until one of the buttons is tapped.

1. Remove this menu (and don't show during this session)
2. Add attachments from other posts (without comments)
3. Add attachments from comments (without other posts)
4. Add ALL attachments of feed (posts and comments)

![Mozilla Firefox 2021-07-31 13-02-00](https://user-images.githubusercontent.com/13445/127736403-07d5a02c-f4b9-4020-b57a-b53e70d06ee6.jpg)

The click on (1) results in item being removed from the loop. User will be able to swipe thru images of set without seeing it again (until she exits lightbox).

The click on (2), (3) or (4) results in the extension of lightbox's dataset:

- (2) All attachments of all currently visible posts (both those which are before and those which are after) are added to the loop and user can swipe thru them.
- (3) "X more comments" is unfolded in the background and all attachments found in current thread are added to the loop.
- (4) All threads on screen are unfolded and attachments found in them are added to the loop.

In all of the cases current position is maintained, number on counter reflects the change. For example, was: 3/3, new 2/72 (it was a first post, html-item is removed) or 43/56 (it was 15-th post), etc.

# Developer-oriented explanation

Implementation of this proposal requires dynamic backend of LightBox. Fortunately, PhotoSwipe v5 (currently in beta) provides good means for this.

- [Dynamically generated data](https://photoswipe.com/v5/docs/data-sources/#dynamically-generated-data): we can change number of items and contents of item
- [Custom HTML-markup](https://photoswipe.com/v5/docs/data-sources/#custom-html-markup) to implement screen with buttons

**Comment by atarka** [comment]
We can also use lightbox navigational header to inform user about current image navigation mode and provide an easy way to change this mode. Interface sample with added controls is shown below. 

![image](https://user-images.githubusercontent.com/10743200/128227022-0e1e77db-a149-4ad6-9e96-b11c4f130158.png)

A button with an icon and a short descriptive text is added. It shows the current viewing mode, it cycles through three possible settings on press. The options are "single post images", "all posts images" and "all images". This setting is stored on client and is persistent between lightbox sessions. 

If something other than "single post images" is selected, a second navigational counter appears that shows currently displayed image's position within the whole image collection (all posts' images or all images on page depending on the selection). This way user can perceive both the current position of an image within a single post and a range of all available images. This is optional, we can use single existing counter that would simply change to reflect selected image navigation mode, but I feel two counters could be more informative.

**Reply by kukutz** [reply]
@atarka great proposal. 
@indeyets do you agree with @atarka idea?

**Reply by indeyets** [reply]
@kukutz I do.

**Reply by kukutz** [reply]
@atarka Can I count on you to do it? :)

**Reply by indeyets** [reply]
@kukutz we need to define actual technical requirements here. And it depends on migrating to newer photoswipe library

I believe it would be appropriate to do migration as a separate first step and then proceeding with this rfc

**Reply by atarka** [reply]
@kukutz I'd be happy to do it.
@indeyets Do we have an estimation on when we are migrating to a newer version?
